Fix/continuous scan performance regression

bparth24 opened the pull request:
# Restore frame-accurate scanning performance

## Summary

Restores 12-16 fps frame-accurate scanning using `requestVideoFrameCallback()` that was lost during the Vue component architecture refactor. Improves performance by 30x-40x over the polling approach.

For details, refer to the README.md and CHANGELOG.md files.

@dlongley cc: @mandyvenables @jameseaster @tminard @msporny @BigBlueHat

bparth24 committed:
Backup scanContinuous before frame-accurate refactor.

- Add BACKUP file with original polling implementation.
- Add REFACTOR-NOTES with objectives and impact analysis.
- Prepare for performance optimization to restore frame callbacks.

bparth24 committed:
Update documentation for enhancedPdf417Plugin

bparth24 committed:
Add frame-accurate scanning method to OpticalScanner.

- Implement _scanContinuousFrameCallback() using requestVideoFrameCallback.
- Provides 30-60 fps scanning vs 0.4 fps polling approach.
- Includes proper timeout and abort signal handling.
- Adds frame count logging for debugging.
- Retries on errors except AbortError.

bparth24 committed:
Add polling fallback method to OpticalScanner.

- Implement _scanContinuousPolling() as compatibility fallback.
- Uses setTimeout with 2.5s intervals for non-video sources.
- Maintains existing polling behavior from original implementation.
- Provides universal compatibility with any source type.
- Will serve as graceful fallback when frame callbacks unavailable.

bparth24 committed:
Refactor scanContinuous to route by source type.

- Replace polling-only implementation with intelligent routing.
- Delegate to _scanContinuousFrameCallback for HTMLVideoElement (fast).
- Delegate to _scanContinuousPolling for other sources (compatible).
- Add warning when non-optimal source type is used.
- Reduce public method complexity from 60+ to 15 lines.
- Keep original implementation as scanContinuous_BACKUP for reference.
- Move backup inline instead of separate file for easier comparison.
- Delete BACKUP-scanContinuous-before-refactor.js file.

bparth24 committed:
Move down backup function

bparth24 committed:
Add comprehensive architecture documentation for continuous scanning.

- Add block comment explaining frame-accurate vs polling strategies.
- Document performance characteristics and trade-offs.
- Include routing logic visualization with ASCII diagram.
- Explain design rationale and historical context.
- Add implementation status section to REFACTOR-NOTES.md.
- Document performance impact and testing requirements.
- Improve maintainability for future developers.

bparth24 committed:
Fix _createScanError helper to preserve error message context.

- Add customMessage parameter to override base error message.
- Add reason parameter for timeout vs user cancellation distinction.
- Update _scanContinuousFrameCallback to use helper with custom messages.
- Update _scanContinuousPolling to use helper with custom messages.
- Ensure consistent error format: 'after N frames/attempts (reason)'.
- Helper now properly preserves original descriptive error messages.

bparth24 committed:
Complete testing and verification of frame-accurate scanning.

- Verify all methods exist and are properly integrated.
- Confirm QR and PDF417 scanning performance improvement.
- Validate MRZ and file upload paths unchanged.
- Test timeout and abort error handling with context.
- Verify frame-accurate scanning achieves 12-16 fps.

bparth24 committed:
Add performance documentation to README.

- Add Performance section highlighting frame-accurate scanning.
- Document scanning strategies (frame-accurate vs polling fallback).
- Include performance metrics table showing 30x-40x improvement.
- Add architecture section explaining continuous scanning implementation.
- Enhance Features list with performance and error context improvements.

This documents the significant performance improvements from the frame-accurate
scanning refactor.

bparth24 committed:
Document frame-accurate scanning refactor in CHANGELOG.

- Add version 2.0.0 entry documenting performance improvements.
- Document 30-40x performance improvement (0.4 fps to 12-16 fps).
- List all new features (frame-accurate scanning, enhanced errors).
- Document refactored scanContinuous method implementation.

This captures the complete scope of the scanning performance optimization.

bparth24 committed:
Update refactor notes

bparth24 committed:
Correct lint issues and update comments

bparth24 committed:
Correct comments in documentation

msporny commented on README.md:
````diff
@@ -150,6 +211,106 @@ Vue (UI Only) -> CameraScanner (Business Logic) -> OpticalScanner (Core Engine)
+```javascript
+async scanContinuous(source, options) {
+  // Check source type
+  if (source instanceof HTMLVideoElement) {
````
```suggestion
  if(source instanceof HTMLVideoElement) {
```

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/6255e6e3efc1c8227a5f7624104234e942e687aa

msporny commented on README.md:
```diff
@@ -150,6 +211,106 @@ Vue (UI Only) -> CameraScanner (Business Logic) -> OpticalScanner (Core Engine)
+// Uses setTimeout with 2.5 second delays
+while (!aborted) {
+  const results = await this.scan(source, options);
+  if (results.length > 0) {
```
```suggestion
while(!aborted) {
  const results = await this.scan(source, options);
  if(results.length > 0) {
```

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/6255e6e3efc1c8227a5f7624104234e942e687aa

msporny commented on README.md:
```diff
@@ -150,6 +211,106 @@ Vue (UI Only) -> CameraScanner (Business Logic) -> OpticalScanner (Core Engine)
+    return this._scanContinuousFrameCallback(source, options);
+  }
+  // Fallback: 0.4 fps polling
+  console.warn('Using polling fallback (slower)');
```
```suggestion
    console.log('bedrock-web-optical-scanner: Using frame-callback (optimal)');
    return this._scanContinuousFrameCallback(source, options);
  }
  // Fallback: 0.4 fps polling
  console.warn('bedrock-web-optical-scanner: Using polling fallback (slower)');
```

Note where the message is coming from. I'm also unsure if we want to print stuff to console if best route is taken... might only want to keep the warning. Keep the console quiet if everything is going well.

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/6255e6e3efc1c8227a5f7624104234e942e687aa

msporny commented on REFACTOR-NOTES.md:
```diff
@@ -0,0 +1,103 @@
+# scanContinuous Refactor - Frame-Accurate Scanning
```
Delete this file, we don't need to put these notes in the repository. You can move these notes to the PR if you want to put them somewhere.

bparth24 replied:
@msporny 

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/df9d87035b1e282df576b69c165a818859169b6c

msporny commented on lib/camera-scanner.js:
```diff
@@ -44,12 +46,22 @@ export class CameraScanner extends EventEmitter {
+    // Validate formats if provided
+    if(formats !== null) {
+      const validFormats = ['qr_code', 'pdf417', 'pdf417_enhanced', 'mrz'];
+      const invalidFormats = formats.filter(f => !validFormats.includes(f));
```
```suggestion
      const invalidFormats = formats.filter(format => !validFormats.includes(format));
```

Avoid using single character variable names unless they are throw-away counters. It causes people reviewing your code, and future-you, to have to expend more effort to understand what each variable refers to.

bparth24 replied:
addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+  * and routes to the appropriate implementation:
+  *
+  *  scanContinuous(source) >
+  *  if (source instanceof HTMLVideoElement) >
```
```suggestion
  *  if(source instanceof HTMLVideoElement) >
```

bparth24 replied:
addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+    if(video instanceof HTMLVideoElement) {
+      // Fast path: Frame-accurate scanning for video elements.
+      console.log('Frame-callback path: Using requestVideoFrameCallback()' +
+        'for optimal performance');
```
```suggestion
```

In general, don't dirty the console if everything is going well.

bparth24 replied:
addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+      '  Recommendation: Use HTMLVideoElement for 30x-40x better performance ' +
+      '(12-16 fps)'
+    );
+    console.log('Polling fallback: Source is not video element, using' +
```
```suggestion
    console.warn('Polling fallback: Source is not video element, using' +
```

bparth24 replied:
addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+      timeout: timeoutMs === 0 ? 'No timeout' : `${timeoutMs}ms`,
+      hasSignal: !!signal
+    });
+
```
```suggestion
```

In general, don't dirty the console w/ debug information. Create a debug flag that is passed into the component or something else for that purpose.

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

logging functionality added in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/3dab62ed94d891d868ac7f7baefa67ebbaf5e5e9

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+            const currentFps = Math.round(1000 / frameDelta);
+            console.log(
+              `Frame ${frameCount}: ${frameDelta}ms delta (~${currentFps} fps)`
+            );
```
Same "don't dirty the console" feedback as above. Move to debug flag or something similar if you want to log performance data.

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

logging functionality added in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/3dab62ed94d891d868ac7f7baefa67ebbaf5e5e9

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+              `delta=${frameDelta}ms, ` +
+              `fps=~${videoFps}` +
+              `${pluginScanDuration > 50 ? 'SLOW' : ''}`
+            );
```
Only log metrics in debug mode.

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

logging functionality added in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/3dab62ed94d891d868ac7f7baefa67ebbaf5e5e9

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
   * @returns {Promise<object[]>} Results when found.
   */
-  async scanContinuous(video, options = {}) {
+  async scanContinuous_BACKUP(video, options = {}) {
```
Don't keep this stuff around if it is no longer needed.

bparth24 replied:
addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

msporny commented on lib/optical-scanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+    // console.log('timeoutMs:', timeoutMs);
+    // console.log('signal provided:', !!signal);
+    // console.log('signal aborted:', signal?.aborted);
+    // console.log('===================================================');
```
Probably need a debug flag for these -- commenting out debug code is a bad code smell -- if you think you might need it in the future, put in a debug capability into the library.

bparth24 replied:
@msporny 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

logging functionality added in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/3dab62ed94d891d868ac7f7baefa67ebbaf5e5e9

msporny commented on lib/plugins/enhancedPdf417Plugin.js:
```diff
@@ -131,6 +136,68 @@ function parseDLInfo(text) {
 
+/**
+ * NOTE: Function Evolution & Changes.
+ * ====================================
+ *
+ * ORIGINAL IMPLEMENTATION (Reference File):
+ * - Spread across multiple Vue component computed properties.
+ * - Used document.body dimensions (clientWidth/clientHeight).
+ * - Applied Math.floor() to regionMaskEdgeLength.
+ * - No bounds checking on calculated values.
+ * - Tightly coupled to Vue component lifecycle.
+ *
+ * CHANGES (Current Implementation):
+ *
+ * 1. DIMENSION SOURCE CHANGE:
+ *  - BEFORE: Used document.body.clientWidth/clientHeight.
+ *  - AFTER: Extracts dimensions from source element itself.
+ *  - IMPACT: More accurate region calculation based on actual media dimensions.
+ *  - CONSIDERATION: Behavior may differ if video size ≠ viewport size.
+ *
+ * 2. CONSOLIDATION:
+ *  - BEFORE: Scattered across computed properties (regionMaskEdgeLength,
+ *    regionLeft, regionTop, region).
+ *  - AFTER: Single, self-contained utility function.
+ *  - BENEFIT: Improved reusability and testability.
```
This has a very strong "generated with an LLM" smell to it -- was it generated using an LLM?

bparth24 replied:
@msporny 

I use Grammarly often to check tone, clarity, and correct grammar while inline documenting and or readme instructions in general.

My approach usually goes: I rough draft with bullet points and then use Grammarly to get tone, clarity, and grammar check.

msporny commented on package-lock.json:
```diff
@@ -9,7 +9,8 @@
         "barcode-detector": "^3.0.5",
-        "dynamsoft-javascript-barcode": "^9.6.1"
+        "dynamsoft-javascript-barcode": "=9.6.1",
+        "dynamsoft-mrz-scanner": "^3.0.1"
```
Don't check in package-lock for libraries.

bparth24 replied:
@msporny 

addressed in this commits - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/4184c6773e97df1b63a947db7f78eba36c086a01 & https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/81dd2e5ff42c2cf991ac036f40529a0e154d6feb

dlongley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,50 @@
 
+## 2.0.0 - 2025-mm-dd
+
+### Added
+
```
```suggestion
### Added
```

dlongley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,50 @@
+  - Detailed comments explaining continuous scanning strategies
+  - Performance characteristics for each approach
+
+### Changed
+
```
```suggestion
### Changed
```

dlongley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,50 @@
+
+### Improved
+
+- Continuous scanning performance: 0.4 fps > 12-16 fps (30x-40X improvement)
```
We don't have an `Improved` section in our changelogs, so we can fold this into Changed`:

```suggestion
- Continuous scanning performance: 0.4 fps > 12-16 fps (30x-40X improvement)
```

bparth24 replied:
@dlongley 

addressed above in commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/114699653fb74ac5bd637de1e65da2993be14f81

dlongley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,50 @@
+  - Provides abort reason (timeout vs user cancellation)
+- Comprehensive architecture documentation in code
+  - Detailed comments explaining continuous scanning strategies
+  - Performance characteristics for each approach
```
Add periods/full stops to the end of all of these to be consistent with our pattern in other libraries.

bparth24 replied:
@dlongley 

addressed above in commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/114699653fb74ac5bd637de1e65da2993be14f81

dlongley commented on CHANGELOG.md:
```diff
@@ -43,4 +88,3 @@
 
```
Should clean up the format of entries in the same ways mentioned above.

bparth24 replied:
@dlongley 

addressed above in commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/114699653fb74ac5bd637de1e65da2993be14f81

dlongley commented on lib/camera-scanner.js:
```diff
@@ -96,20 +108,30 @@ export class CameraScanner extends EventEmitter {
   _getFormats(scanType) {
+    // Check for explicit format override first
+    if(this.config.formats !== null) {
+      return this.config.formats; // returns ['pdf417_enhanced']
```
Put comments above to match our coding style... but does this comment still apply?

```suggestion
      // returns ['pdf417_enhanced']
      return this.config.formats;
```

bparth24 replied:
@dlongley 

Noted. Yes, this comment is still valid.

However, I have explained here ([link](https://github.com/digitalbazaar/bedrock-vue-optical-scanner/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R75-R82)) that the use is primarily to support legacy production application that uses the Dynamsoft SDK for pdf417 scan.

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/dcfcb5cae4a7d8aa90bbdbe1a4ca62165d2fbbd9

dlongley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,50 @@
+- Debugging: Rich error messages with frame/attempt counts
+
+### Fixed
+
```
```suggestion
### Fixed
```

dlongley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,50 @@
+
+- Performance regression from original `bedrock-vue-barcode-scanner` implementation
+  - Restored frame-accurate scanning that was lost during refactor
+  - Now matches or exceeds original performance
```
This changelog says it's a major version bump, but none of these items use the prefix `**BREAKING**: `. Please update the breaking items with that prefix so readers can find them quickly. If there are actually no breaking API changes, we should make this a minor update instead.

bparth24 replied:
@dlongley 

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/c2c19e9e6b60dbaf7310f442e10f00f5fb42be0a

I would like to point out, no breaking changes if developers are using `bedrock-web-optical-scanner`, so it will be a minor update. If developers are migrating from `barcode-vue-barcode-scanner`, then there might be a difference in how to use `bedrock-vue-optical-scanner`, as our approach is to separate the UI concerns and scanning logic.

I hope that makes sense.

dlongley commented on README.md:
````diff
@@ -12,12 +12,68 @@ allows you to easily add support for new formats and scanning modes.
+
+```javascript
+// Automatic routing - no configuration needed
+if (source instanceof HTMLVideoElement) {
````
```suggestion
if(source instanceof HTMLVideoElement) {
```

bparth24 replied:
@dlongley 

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/6255e6e3efc1c8227a5f7624104234e942e687aa

dlongley commented on lib/plugins/mrzPlugin.js:
```diff
@@ -126,9 +126,11 @@ export const mrzPlugin = {
+      {
+        licenseKey: !!licenseKey,
+        scannerConfig, targetContainer: !!scannerConfig.targetContainer
+      });
```
We probably don't want to print this license key to the console. Dynamsoft's license key mechanism leaves much to be desired, but we don't need to surface it this easily.

bparth24 replied:
@dlongley 

Added debug flag for console logs. It will not be printed on console in production with `debug=false` config.

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2a8110dacdaeb265450ce78a8557a58d8d1d9297

dlongley commented on lib/opticalScanner.js:
```diff
@@ -78,12 +78,14 @@ export class OpticalScanner {
       const plugin = this.plugins.get(format);
```
The file naming format should be consistent (it's kebab case here and camel case in `plugins`). It should use camel case naming in both places, so change the kebab case ones to: `opticalScanner.js` and `cameraScanner.js`.

bparth24 replied:
@dlongley 

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/dcfcb5cae4a7d8aa90bbdbe1a4ca62165d2fbbd9

dlongley commented on lib/opticalScanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+
+          // No results yet - schedule next frame.
+          video.requestVideoFrameCallback(scanFrame);
+
```
```suggestion
```

dlongley commented on lib/opticalScanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+
+    // Create timeout-aware abort signal.
+    const controller = this._createTimeoutController(signal, timeoutMs);
+    const effectiveSignal = controller?.signal || signal;
```
It seems like this could be reworked to just produce a variable named `signal` ... instead of calling the one that is used everywhere `effective`. A quick scan doesn't show `controller` being used anywhere other than for its signal ... and it looks like `_createTimeoutController` is meant to be some kind of combiner on signals. It was hard to follow what it's doing, but could it be simplified by using the built-in signal combiner?

https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/any_static

Perhaps `AbortSignal.timeout()` might also be of some use here?

bparth24 replied:
@dlongley 

I understand. I will simplify and improve the code structure/flow.

**Context:** Related to signal and timeout. I was trying to achieve two things in one go —
1. User cancellation signal (passed from Vue) 
2. Automatic timeout signal (based on timeout param)

Additionally, for timeout values, based on testing results, I have set default values here:
https://github.com/digitalbazaar/bedrock-web-optical-scanner/blob/fix/continuous-scan-performance-regression/lib/cameraScanner.js#L261C3-L295


**Note:** Thinking out loud here, and have not tested the code below. Here is what I'm thinking of implementing instead of `_createTimeoutController()`. Before I implement it, I would like to hear your feedback.

```
_combineSignals(userSignal, timeoutMs) {
  // No signal or timeout needed
  if (!userSignal && timeoutMs <= 0) {
    return undefined;
  }

  // Collect signals to combine
  const signals = [];
  if (userSignal) signals.push(userSignal);
  if (timeoutMs > 0) signals.push(AbortSignal.timeout(timeoutMs));

  // Return combined signal (or single signal if only one)
  return signals.length === 1 ? signals[0] : AbortSignal.any(signals);
} 
```

dlongley replied:
+1 Yeah, that looks cleaner (modulo style adjustments of course).

bparth24 replied:
@dlongley, appreciate the quick feedback.

bparth24 replied:
> It seems like this could be reworked to just produce a variable named `signal` ... instead of calling the one that is used everywhere `effective`. A quick scan doesn't show `controller` being used anywhere other than for its signal ... and it looks like `_createTimeoutController` is meant to be some kind of combiner on signals. It was hard to follow what it's doing, but could it be simplified by using the built-in signal combiner?
> 
> https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/any_static
> 
> Perhaps `AbortSignal.timeout()` might also be of some use here?

@dlongley 

Addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/99c32ef32dc156b6f3d5dc100a8e07d55ea3841e and https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2b84580118af5b1c6f666da0ca0e5bf832fde3d3.

dlongley commented on lib/opticalScanner.js:
```diff
@@ -128,22 +130,395 @@ export class OpticalScanner {
+      let lastFrameTime = Date.now();
+      let frameCount = 0;
+
+      const scanFrame = async () => {
```
I suspect this could be reworked to just call an async function instead -- which will improve maintenance. This scheduling function seems to be doing a lot (maybe it's mostly metric calculations though?) and a signal to me that it could be refactored to be simpler is the fact that `video.requestVideoFrameCallback(scanFrame);` is called in three independent places. My intuition is that those could be coerced into a single code path, resulting in something that's simpler to follow and maintain.

bparth24 replied:
> I suspect this could be reworked to just call an async function instead -- which will improve maintenance. This scheduling function seems to be doing a lot (maybe it's mostly metric calculations though?) and a signal to me that it could be refactored to be simpler is the fact that `video.requestVideoFrameCallback(scanFrame);` is called in three independent places. My intuition is that those could be coerced into a single code path, resulting in something that's simpler to follow and maintain.

@dlongley 

The attempt to address the above concerns in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/2e61acea9df5e52304014b35229e91716ae5813c.

bparth24 replied:
@dlongley 

Additionally, I will improve the existing coding pattern to use async functions instead of new Promise(...).

bparth24 committed:
Update .gitignore to ignore package-lock.json

bparth24 committed:
Delete Refactor-notes.md

bparth24 committed:
Update readme - correct formatting in code

bparth24 committed:
Update changelog - changed section, remove improved section.

bparth24 committed:
Add logging functionality

bparth24 committed:
Update console log with proper debug flag - logging functionality.

bparth24 committed:
Remove package-lock.json

bparth24 committed:
Rename files with camel case format, fix inline comments to top (separate) line comments.

bparth24 committed:
Correct file name in readme and version in changelog file

davidlehn commented on lib/plugins/enhancedPdf417Plugin.js:
```diff
@@ -7,6 +7,9 @@ import {BarcodeDetector} from 'barcode-detector/ponyfill';
 
 /**
 * Driver License field mappings for PDF417 data parsing.
+* Note: This code is exactly similar to bedrock-bue-pdf417/lib/helpers/
```
```suggestion
* Note: This code is exactly similar to bedrock-vue-pdf417/lib/helpers/
```

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/enhancedPdf417Plugin.js:
```diff
@@ -98,6 +101,8 @@ const driverLicenseFields = {
 
 /**
 * Parse driver license data from PDF417 text.
+* Note: This function exactly similar to bedrock-bue-pdf417/BarcodeScanner.vue
```
```suggestion
* Note: This function exactly similar to bedrock-vue-pdf417/BarcodeScanner.vue
```

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/enhancedPdf417Plugin.js:
```diff
@@ -7,6 +7,9 @@ import {BarcodeDetector} from 'barcode-detector/ponyfill';
 
 /**
 * Driver License field mappings for PDF417 data parsing.
+* Note: This code is exactly similar to bedrock-bue-pdf417/lib/helpers/
```
- "exactly similar to" is an odd phrase.  A bit too zen for docs.  Maybe say "the same as"
- Same comment for getDLInfo below.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/enhancedPdf417Plugin.js:
```diff
@@ -198,22 +265,22 @@ export const enhancedPdf417Plugin = {
-   *
-   * @returns {Promise<object[]>} Array of detected PDF417 codes.
-   */
+  * Scan source for PDF417 codes using enhanced algorithms.
```
Some inconsistencies in these files.  Usual indent style is to align first asterisk as what was here before:
```js
/**
 * xxx
 */
async foo() ...
```

bparth24 replied:
@davidlehn 

Understood your concern. I was trying to either indent with 2 spaces or no space and avoid having 1 space before `*`. 

```
/**
 * xxx
 */
async foo() ...
```

What's the correct styling we follow for other libraries?

davidlehn replied:
I'm pretty sure everyone uses the align first asterisk style.  Look at any of our libs or the jsdoc docs.

bparth24 replied:
@davidlehn 

Noted. I will correct in all files.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/enhancedPdf417Plugin.js:
```diff
@@ -198,22 +265,22 @@ export const enhancedPdf417Plugin = {
+  * @param {object} options - Plugin-specific options.
+  * @param {AbortSignal} options.signal - Abort signal.
+  * @param {string} options.license - Dynamsoft license key.
+  * @param {boolean} options.useDynamsoft - Use Dynamsoft engine.
```
Use markup for all the optional params here:
```suggestion
  * @param {boolean} [options.useDynamsoft] - Use Dynamsoft engine.
```

bparth24 replied:
@davidlehn 

Noted. I will use `[]` if it is an optional parameter whenever it is applicable.

davidlehn commented on lib/plugins/index.js:
```diff
@@ -2,10 +2,9 @@
  */
 
-// Import everything first
+// Import
```
Is there a reason to add these obvious plain Import/Export comments?

bparth24 replied:
No specific reasons. @davidlehn

I have tendency to start with placeholder comments when I create a new file and then start adding code below those placeholder value.

I have removed.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/mrzPlugin.js:
```diff
@@ -21,20 +23,26 @@ export const mrzPlugin = {
-    console.log('========================');
+    if(isDebugEnabled()) {
+      console.log(
+        ...logger.prefix('=== MRZ PLUGIN DEBUG ==='));
```
Text duplication with `[MRZ Plugin]` prefix.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/mrzPlugin.js:
```diff
@@ -44,8 +52,11 @@ export const mrzPlugin = {
-    // console.log('MRZ Mode resolved to:', mrzMode);
+    if(isDebugEnabled()) {
+      // console.log(
+      //   ...logger.prefix('MRZ Plugin: About to enter switch statement'));
```
Text duplication with `[MRZ Plugin]` prefix.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/mrzPlugin.js:
```diff
@@ -64,11 +75,13 @@ export const mrzPlugin = {
         case 'camera':
-          console.log('MRZ Plugin: Entering camera mode');
+          isDebugEnabled() &&
+            console.log(...logger.prefix('MRZ Plugin: Entering camera mode'));
```
Text duplication with `[MRZ Plugin]` prefix.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on lib/plugins/mrzPlugin.js:
```diff
@@ -64,11 +75,13 @@ export const mrzPlugin = {
           );
-          console.log('MRZ Plugin: Camera scan completed:', result);
+          isDebugEnabled() && console.log(
+            ...logger.prefix('MRZ Plugin: Camera scan completed:', result));
```
Text duplication with `[MRZ Plugin]` prefix.

bparth24 replied:
@davidlehn 

addressed in this commit - https://github.com/digitalbazaar/bedrock-web-optical-scanner/pull/5/commits/8460c32c53c28fec1c7c87fc64a948c0aca2ca21

davidlehn commented on CHANGELOG.md:
```diff
@@ -1,5 +1,30 @@
+## 1.1.0 - 2025-mm-dd
+
+### Added
+
```
Normal style is no blank line after these sub headers.
```suggestion
```

bparth24 committed:
Cleanup, format - documentation section and comments

bparth24 committed:
Use browser APIs for signal handling.

- Replaces custom _createTimeoutController() with _combineSignals().
- Uses AbortSignal.any() and AbortSignal.timeout().
- Eliminates manual cleanup.
- Renames effectiveSignal to signal for clarity.

bparth24 committed:
Remove _createTimeoutController function.

bparth24 committed:
Validate license before MRZ scanning.

- Checks for Dynamsoft license key in CameraScanner.scan().
- Throws clear error.

bparth24 committed:
Refactor frame loop and simplify documentation.

Frame Loop Refactoring:
- Extract _runFrameLoop() for separation of concerns.
- Replace _createTimeoutController() with _combineSignals().
- Use standard AbortSignal.any() and AbortSignal.timeout() APIs.
- Extract metrics and logging helpers for maintainability.

Documentation Improvements:
- Reduce inline documentation verbosity in opticalScanner.js.
- Fix performance metric inconsistencies (62-83ms throughout).
- Add device specifications to README (MacBook Pro M1, Chrome).
- Remove duplicate architecture details from method comments.

Addresses code review feedback on maintainability and clarity.
